### PR TITLE
Add retry logic for proxy's search and query tasks

### PR DESCRIPTION
**Pull request by letian-jiang**
Related to #17098
Retry GetShardLeaders until service available or timeout

Signed-off-by: Letian Jiang <letian.jiang@zilliz.com>

### CODE REVIEW

**Comment by letian-jiang** [comment]
Work in progress. /hold

**Comment by letian-jiang** [comment]
/hold 
Work in progress.

**Comment by letian-jiang** [comment]
rerun ut

**Review by congqixia — COMMENTED** [review]
Maybe retry shall be applied to `GetShardLeaders` grpc
@XuanYang-cn ptal

File: internal/proxy/meta_cache.go [hunk 1]
```diff
@@ -607,12 +609,23 @@ func (m *MetaCache) GetShards(ctx context.Context, withCache bool, collectionNam
...
+	var resp *querypb.GetShardLeadersResponse
+	childCtx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	err = retry.Do(childCtx, func() error {
+		resp, err = qc.GetShardLeaders(ctx, req)
+		if err != nil {
+			return err
+		}
+		if resp.Status.ErrorCode != commonpb.ErrorCode_Success {
+			return fmt.Errorf("fail to get shard leaders from QueryCoord: %s", resp.Status.Reason)
```
Comment (congqixia) [inline internal/proxy/meta_cache.go:623]: Maybe check the error code. Return `retry.Unrecoverable` in some case other than replica not ready.

File: internal/proxy/meta_cache.go [hunk 2]
```diff
@@ -607,12 +609,23 @@ func (m *MetaCache) GetShards(ctx context.Context, withCache bool, collectionNam
...
+		}
+		return nil
+	})
 	if err != nil {
-		return nil, err
-	}
-	if resp.Status.ErrorCode != commonpb.ErrorCode_Success {
-		return nil, fmt.Errorf("fail to get shard leaders from QueryCoord: %s", resp.Status.Reason)
+		return nil, fmt.Errorf("GetShardLeaders timeout")
```
Comment (congqixia) [inline internal/proxy/meta_cache.go:628]: Full list of all retry need to be printed.

**Review by congqixia — COMMENTED** [review]
There are some maybe NIT comment.
@XuanYang-cn ptal

File: internal/proxy/meta_cache.go [hunk 3]
```diff
@@ -607,12 +609,23 @@ func (m *MetaCache) GetShards(ctx context.Context, withCache bool, collectionNam
...
-	resp, err := qc.GetShardLeaders(ctx, req)
+
+	// retry until service available or context timeout
+	var resp *querypb.GetShardLeadersResponse
+	childCtx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	err = retry.Do(childCtx, func() error {
+		resp, err = qc.GetShardLeaders(ctx, req)
+		if err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (letian-jiang) [reply internal/proxy/meta_cache.go]: Fixed. But I am not sure whether we should retry on grpc error

File: internal/proxy/meta_cache.go (hunk 1 above)
Comment (letian-jiang) [reply internal/proxy/meta_cache.go]: Fixed. Add `NoReplicaAvailable` error code.

File: internal/proxy/meta_cache.go (hunk 2 above)
Comment (letian-jiang) [reply internal/proxy/meta_cache.go]: Fixed.

**Review by congqixia — COMMENTED** [review]
/lgtm

**Pull request merged**